vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,363 @@
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformClose()
+{
+  co_await resume_background();
+
+  m_socket.Close(static_cast<uint16_t>(m_closeCode), Utf8ToUtf16(m_closeReason));
+
+  co_return;
+}
+
+void WinRTWebSocketResource::OnMessageReceived(IWebSocket const& sender, MessageWebSocketMessageReceivedEventArgs const& args)
+{
+  try
+  {
+    string response;
+    DataReader reader = args.GetDataReader();
+    auto len = reader.UnconsumedBufferLength();
+    if (args.MessageType() == SocketMessageType::Utf8)
+    {
+      reader.UnicodeEncoding(UnicodeEncoding::Utf8);
+      vector<uint8_t> data(len);
+      reader.ReadBytes(data);
+
+      response = string(CheckedReinterpretCast<char*>(data.data()), data.size());
+    }
+    else
+    {
+      auto buffer = reader.ReadBuffer(len);
+      winrt::hstring data = CryptographicBuffer::EncodeToBase64String(buffer);
+
+      response = Utf16ToUtf8(std::wstring_view(data));
+    }
+
+    if (m_readHandler)
+      m_readHandler(response.length(), response);//TODO: move?
+  }
+  catch (hresult_error const& e)
+  {
+    //TODO: Re-enable (fix "OutputDebugStringW': function does not take 2 arguments")
+    //OutputDebugString("Read failed (0x%8X) %ls\n", e);
+    m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Receive });
+  }
+}
+
+void WinRTWebSocketResource::OnClosed(IWebSocket const& sender, WebSocketClosedEventArgs const& args)
+{
+  if (m_closeHandler)
+  {
+    //TODO: Parameterize (pass via member variables?)
+    m_closeHandler(CloseCode::Normal, "Closing");
+  }
+}
+
+void WinRTWebSocketResource::Stop()
+{
+  // Ensure sequence of other operations
+  if (m_connectRequested)
+  {
+    // m_connectPerformed.Wait();//TODO: Set timeout?
+    m_connectPerformed.get_future().wait();
+  }
+}
+
+#pragma endregion Private members
+
+#pragma region IWebSocketResource
+
+void WinRTWebSocketResource::Connect(const Protocols& protocols, const Options& options)
+{
+  for (auto& header : options)
+  {
+    m_socket.SetRequestHeader(header.first, Utf8ToUtf16(header.second));
+  }
+
+  winrt::Windows::Foundation::Collections::IVector<winrt::hstring> supportedProtocols =
+    m_socket.Control().SupportedProtocols();
+  for (auto& protocol : protocols)
+  {
+    supportedProtocols.Append(Utf8ToUtf16(protocol));
+  }
+
+  m_connectRequested = true;
+  PerformConnect();
+}
+
+void WinRTWebSocketResource::Ping()
+{
+  PerformPing();
+}
+
+void WinRTWebSocketResource::Send(const string& message)
+{
+  m_writeQueue.emplace(std::move(message), false);
+
+  PerformWrite();
+  //PerformWrite(std::move(message));
+}
+
+void WinRTWebSocketResource::SendBinary(const string& base64String)
+{
+  m_writeQueue.emplace(std::move(base64String), true);
+
+  PerformWrite();
+}
+
+void WinRTWebSocketResource::Close(CloseCode code, const string& reason)
+{
+  Stop();
+
+  m_closeCode = code;
+  m_closeReason = std::move(reason);
```

Comment:
`std::move` on something const is useless, we'll just do the copy.

Reply:
Correct!

In a future iteration, I'll likely remove the `const&` qualifiers from the interface.
They don't really make sense for this API's purpose.